### PR TITLE
[GEP-31] Controller optimizations and Add e2e tests

### PR DESCRIPTION
**Pull request by shafeeqes**
**How to categorize this PR?**

/area control-plane delivery open-source scalability
/kind enhancement

**What this PR does / why we need it**:
- This PR adds optimization to some of the controllers for handling edge cases
- Add e2e test for in-place credentials rotation
- Add e2e tests for machine image, kubernetes version, and kubelet config updates.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/10219

**Special notes for your reviewer**:
/hold until https://github.com/gardener/gardener/pull/11963 is merged

**Release note**:

```other operator
NONE
```

### CODE REVIEW

**Comment by shafeeqes** [comment]
/unhold
https://github.com/gardener/gardener/pull/11963 is merged

File: extensions/pkg/controller/worker/controller.go [hunk 1]
```diff
@@ -162,8 +163,8 @@ func MachineConditionChangedPredicate(ctx context.Context, log logr.Logger, c cl
 			oldCond := GetMachineCondition(oldMachine, machinev1alpha1.NodeInPlaceUpdate)
 			newCond := GetMachineCondition(newMachine, machinev1alpha1.NodeInPlaceUpdate)
 
-			// Consider only the condition transition from UpdateCandidate to SelectedForUpdate
-			return oldCond != nil && newCond != nil && oldCond.Reason == machinev1alpha1.CandidateForUpdate && newCond.Reason == machinev1alpha1.SelectedForUpdate
+			// Consider only the condition transition from CandidateForUpdate to SelectedForUpdate
```
Comment (acumino) [inline extensions/pkg/controller/worker/controller.go]:
```suggestion
			// Consider only the condition transition from CandidateForUpdate to something else
```

File: pkg/apiserver/registry/core/shoot/strategy.go [hunk 2]
```diff
@@ -178,6 +178,8 @@ func (shootStrategy) Validate(_ context.Context, obj runtime.Object) field.Error
 	allErrs = append(allErrs, validation.ValidateShoot(shoot)...)
 	allErrs = append(allErrs, validation.ValidateForceDeletion(shoot, nil)...)
 	allErrs = append(allErrs, validation.ValidateFinalizersOnCreation(shoot.Finalizers, field.NewPath("metadata", "finalizers"))...)
+	allErrs = append(allErrs, validation.ValidateInPlaceUpdateStrategyOnCreation(shoot)...)
```
Comment (acumino) [inline pkg/apiserver/registry/core/shoot/strategy.go:181]: So, if a user updates the shoot and adds a new worker pool with in-place update enabled, it will be allowed. Isn't that incorrect? Ideally, it should only be allowed if the existing worker pool already has in-place update enabled — otherwise, it should throw an error for the new worker pool.

File: test/e2e/gardener/shoot/shoot.go [hunk 3]
```diff
@@ -265,3 +272,156 @@ func ItShouldAnnotateShoot(s *ShootContext, annotations map[string]string) {
...
+			}
+		}
+	}, SpecTimeout(2*time.Minute))
+}
+
+// ItShouldFindAllMachinePodsBefore finds all machine pods before running the required tests and returns their names.
+func ItShouldFindAllMachinePodsBefore(s *ShootContext) sets.Set[string] {
+	GinkgoHelper()
+
+	beforeStartMachinePodNames := sets.New[string]()
```
Comment (acumino) [inline test/e2e/gardener/shoot/shoot.go]:
```suggestion
	machinePodsBeforeTest := sets.New[string]()
```

File: test/e2e/gardener/shoot/shoot.go [hunk 4]
```diff
@@ -265,3 +272,156 @@ func ItShouldAnnotateShoot(s *ShootContext, annotations map[string]string) {
...
+
+	It("Find all machine pods to ensure later that they weren't rolled out", func(ctx SpecContext) {
```
Comment (acumino) [inline test/e2e/gardener/shoot/shoot.go:309]: same for all its

```suggestion
	It("find all machine pods to ensure later that they weren't rolled out", func(ctx SpecContext) {
```

File: pkg/apiserver/registry/core/shoot/strategy.go (hunk 2 above)
Comment (shafeeqes) [reply pkg/apiserver/registry/core/shoot/strategy.go:181]: Right 🤦‍♂️ , will fix it.

File: test/e2e/gardener/shoot/shoot.go (hunk 4 above)
Comment (shafeeqes) [reply test/e2e/gardener/shoot/shoot.go:309]: We use capital letters in lot of other `It`s as well.

Comment (acumino) [reply test/e2e/gardener/shoot/shoot.go:309]: This is inconsistent, why some of them have capital letter and some don't.

**Review by acumino — COMMENTED** [review]
/lgtm

File: extensions/pkg/controller/worker/controller.go [hunk 5]
```diff
@@ -109,6 +109,7 @@ func MachineConditionChangedPredicate(ctx context.Context, log logr.Logger, c cl
 
 			machineDeploymentName, ok := machine.Labels[LabelKeyMachineDeploymentName]
 			if !ok {
+				log.Error(nil, "Machine does not have machine deployment label", "machine", machine.Name)
```
Comment (rfranzke) [inline extensions/pkg/controller/worker/controller.go]: log.Info?

File: test/e2e/gardener/shoot/shoot.go [hunk 6]
```diff
@@ -265,3 +272,156 @@ func ItShouldAnnotateShoot(s *ShootContext, annotations map[string]string) {
...
+	GinkgoHelper()
+
+	It("should label all the manual in-place nodes with selected-for-update", func(ctx SpecContext) {
+		for _, pool := range s.Shoot.Spec.Provider.Workers {
+			if !v1beta1helper.IsUpdateStrategyManualInPlace(pool.UpdateStrategy) {
+				continue
+			}
+
+			nodeList := &corev1.NodeList{}
+			Eventually(ctx, s.ShootKomega.List(nodeList, client.MatchingLabels{v1beta1constants.LabelWorkerPool: pool.Name})).Should(Succeed())
```
Comment (rfranzke) [inline test/e2e/gardener/shoot/shoot.go]: description should contain pool name

File: test/e2e/gardener/shoot/shoot.go [hunk 7]
```diff
@@ -265,3 +272,156 @@ func ItShouldAnnotateShoot(s *ShootContext, annotations map[string]string) {
...
+
+			for _, node := range nodeList.Items {
+				if metav1.HasLabel(node.ObjectMeta, machinev1alpha1.LabelKeyNodeSelectedForUpdate) {
+					continue
+				}
+
+				Eventually(ctx, s.ShootKomega.Update(&node, func() {
```
Comment (rfranzke) [inline test/e2e/gardener/shoot/shoot.go:295]: description should contain pool name

File: test/e2e/gardener/shoot/shoot.go [hunk 8]
```diff
@@ -265,3 +272,156 @@ func ItShouldAnnotateShoot(s *ShootContext, annotations map[string]string) {
...
+			afterStartMachinePodNames.Insert(item.Name)
+		}
+
+		Expect(beforeStartMachinePodNames.UnsortedList()).To(ConsistOf(afterStartMachinePodNames.UnsortedList()))
+	}, SpecTimeout(time.Minute))
+}
+
+// ItShouldRewriteOsRelease rewrites the /etc/os-release file for all machine pods to ensure that the version is overwritten for tests.
+// This is a workaround for the fact that the machine image version is not available in the os-release file in the local provider.
+func ItShouldRewriteOsRelease(s *ShootContext) {
```
Comment (rfranzke) [inline test/e2e/gardener/shoot/shoot.go]: OS

File: test/e2e/gardener/shoot/shoot.go [hunk 9]
```diff
@@ -265,3 +272,156 @@ func ItShouldAnnotateShoot(s *ShootContext, annotations map[string]string) {
...
+				fmt.Sprintf(
+					`s/^PRETTY_NAME="[^"]*"/PRETTY_NAME="Machine Image Version %s (version overwritten for tests, check VERSION_ID for actual version)"/`,
+					*s.Shoot.Spec.Provider.Workers[poolIndex].Machine.Image.Version,
+				),
+				"/etc/os-release",
+			)
+
+			Expect(err).NotTo(HaveOccurred())
+		}
+	}, SpecTimeout(2*time.Minute))
```
Comment (rfranzke) [inline test/e2e/gardener/shoot/shoot.go]: This is weird. Shouldn't this be done by `provider-local` itself somehow? Doing this in the e2e test feels wrong.

File: pkg/provider-local/controller/operatingsystemconfig/actuator.go [hunk 10]
```diff
@@ -79,13 +78,14 @@ systemctl daemon-reload
 	return operatingsystemconfig.WrapProvisionOSCIntoOneshotScript(script), nil
 }
 
-func (a *actuator) handleReconcileOSC(osc *extensionsv1alpha1.OperatingSystemConfig) ([]extensionsv1alpha1.Unit, []extensionsv1alpha1.File, *extensionsv1alpha1.InPlaceUpdatesStatus, error) {
+func (a *actuator) handleReconcileOSC(osc *extensionsv1alpha1.OperatingSystemConfig) ([]byte, []extensionsv1alpha1.Unit, []extensionsv1alpha1.File, *extensionsv1alpha1.InPlaceUpdatesStatus, error) {
```
Comment (rfranzke) [inline pkg/provider-local/controller/operatingsystemconfig/actuator.go]: This change is weird. The `reconcile` OSC has nothing to do with user data, hence, the function cannot return any meaningful. What is the point in adding this then to the return parameters in the first place?

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: test/e2e/gardener/shoot/shoot.go (hunk 9 above)
Comment (shafeeqes) [reply test/e2e/gardener/shoot/shoot.go:387]: This is supposed to be done by the user in the real scenario, here the e2e test is the "user".
We can also write a controller in provider-local to watch for machines and do this, but this is not needed in any other test other than in-place.

Comment (shafeeqes) [reply test/e2e/gardener/shoot/shoot.go]: Fixed with https://github.com/gardener/gardener/pull/11953/commits/03b9f2aeb2babb312d45b9618cf333f01e4996d5

**Comment by shafeeqes** [comment]
@rfranzke PTAL

**Review by rfranzke — COMMENTED** [review]
Very nice, I like the approach :) Thank you!

/lgtm
